test/snippet/io/sequence_file/sequence_file_input.cpp

```diff
@@ -0,0 +1,216 @@
+//              ^ no need to specify the template arguments
+//! [istringstream]
+(void) fin;
+}
+
+{
+//! [aminoacid]
+sequence_file_input<sequence_file_input_default_traits_aa> fin{"/tmp/my.fasta"};
+//! [aminoacid]
+(void) fin;
+}
+
+{
+std::string input
+{
+    "> TEST1\n"
+    "FQTWE\n"
+    "> Test2\n"
+    "KYRTW\n"
+    "> Test3\n"
+    "EEYQTWEEFARAAEKLYLTDPMKV\n"
+};
+//! [template_specification]
+// ... input had amino acid sequences
+std::istringstream iss(input);
+
+sequence_file_input<sequence_file_input_default_traits_aa,
+                    fields<field::SEQ, field::ID, field::QUAL>,
+                    type_list<sequence_file_format_fasta>,
+                    char> fin{std::move(iss), sequence_file_format_fasta{}};
+//! [template_specification]
+(void) fin;
+}
+
+{
+//! [record_iter]
+sequence_file_input fin{"/tmp/my.fasta"};
+
+for (auto & rec : fin)
+{
+    std::cout << "ID:  " << get<field::ID>(rec) << '\n';
+    std::cout << "SEQ: " << (get<field::SEQ>(rec) | view::to_char) << '\n'; // sequence is converted to char on-the-fly
+    // a quality field also exists, but is not printed, because we know it's empty for FastA files.
+}
+//! [record_iter]
+}
+
+{
+//! [auto_ref]
+sequence_file_input fin{"/tmp/my.fasta"};
+
+using record_type = typename decltype(fin)::record_type;
+std::vector<record_type> records;
+
+for (auto & rec : fin)
+    records.push_back(std::move(rec));
+//! [auto_ref]
+}
+
+{
+//! [decomposed]
+sequence_file_input fin{"/tmp/my.fasta"};
+
+for (auto & [ seq, id, qual ] : fin)
+{
+    std::cout << "ID:  " << id << '\n';
+    std::cout << "SEQ: " << (seq | view::to_char) << '\n'; // sequence is converted to char on-the-fly
+    std::cout << "EMPTY QUAL." << (qual | view::to_char) << '\n';            // qual is empty for FastA files
+}
+//! [decomposed]
+}
+
+{
+//! [custom_fields]
+sequence_file_input fin{"/tmp/my.fasta", fields<field::ID, field::SEQ_QUAL>{}};
+
+for (auto & [ id, seq_qual ] : fin) // note that the order is now different, "id" comes first, because it was specified first
+{
+    std::cout << "ID:  " << id << '\n';
+    // sequence and qualities are part of the same vector, of type std::vector<dna5q>
+    std::cout << "SEQ: "  << (seq_qual | view::get<0> | view::to_char) << '\n'; // sequence string is extracted and converted to char  on-the-fly
+    std::cout << "QUAL: " << (seq_qual | view::get<1> | view::to_char) << '\n'; // quality string is extracted and converted to char  on-the-fly
+}
+//! [custom_fields]
+}
+
+{
+//! [file_view]
+sequence_file_input fin{"/tmp/my.fasta"};
+
+auto minimum_length5_filter = view::filter([] (auto const & rec)
+{
+    return size(get<field::SEQ>(rec)) >= 5;
+});
+
+for (auto & rec : fin | minimum_length5_filter) // only record with sequence length >= 5 will "appear"
+{
+    std::cout << "IDs of seq_length >= 5: " << get<field::ID>(rec) << '\n';
+    // ...
+}
+//! [file_view]
+}
+
+{
+#if 0
```

Comment:
The struct itself has to be defined outside of main. So I define the struct and test the lines of code outside of the snippet itself, and then just do this so that all the lines are in one place.